### PR TITLE
SONAR-10040 Declare constraints on Web Service parameters

### PR DESCRIPTION
**Pull request by gjambet**
QA is failing as it requires changes on UI side

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ListAction.java#L133](https://github.com/SonarSource/sonarqube/blob/f48a35afa8d35a1666a23bc05ae42c5c8a94aa81/server/sonar-server/src/main/java/org/sonar/server/branch/ws/ListAction.java#L133): Specify a "BranchDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ListAction.java#L133](https://github.com/SonarSource/sonarqube/blob/193951fefdae5bf79545fbedf4d036e6089f0d08/server/sonar-server/src/main/java/org/sonar/server/branch/ws/ListAction.java#L133): Specify a "BranchDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ListAction.java#L133](https://github.com/SonarSource/sonarqube/blob/a43092be57c1a010cd8c07fb911b6c9e351fd176/server/sonar-server/src/main/java/org/sonar/server/branch/ws/ListAction.java#L133): Specify a "BranchDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ListAction.java#L133](https://github.com/SonarSource/sonarqube/blob/d9a9a0788040e59ef59278c53d9ac98bf876c7e9/server/sonar-server/src/main/java/org/sonar/server/branch/ws/ListAction.java#L133): Specify a "BranchDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ListAction.java#L133](https://github.com/SonarSource/sonarqube/blob/c9b525e7b964ff2bf2bddd7ae524225bab562601/server/sonar-server/src/main/java/org/sonar/server/branch/ws/ListAction.java#L133): Specify a "BranchDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ListAction.java#L133](https://github.com/SonarSource/sonarqube/blob/3050b68fbcd2d87a5a866b441632aed47ee21b48/server/sonar-server/src/main/java/org/sonar/server/branch/ws/ListAction.java#L133): Specify a "BranchDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGatesWs.java [hunk 1]
```diff
@@ -78,10 +80,12 @@ static void addConditionParams(NewAction action) {
       .setPossibleValues("1");
 
     action.createParam(PARAM_WARNING)
+      .setMaximumLength(VALUE_LENGTH)
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGatesWs.java]: rename to CONDITION_MAX_LENGTH

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java [hunk 2]
```diff
@@ -73,23 +77,29 @@ public void define(WebService.NewController context) {
...
+      .setMaximumLength(MAX_NAME_LENGTH)
+      .setDescription("Name")
+      .setExampleValue("Team Size");
+
     action.createParam(PARAM_TYPE)
-      .setDescription("Type")
-      .setPossibleValues(Metric.ValueType.names());
+      .setDescription("Metric type key")
+      .setPossibleValues(Metric.ValueType.names())
+      .setExampleValue(Metric.ValueType.INT.name());
```
[the remark this answers is not in the document]
Comment (gjambet) [reply server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java]: done

File: server/sonar-server/src/test/java/org/sonar/server/project/ws/CreateActionTest.java [hunk 3]
```diff
@@ -48,7 +47,8 @@
 import org.sonarqube.ws.WsProjects.CreateWsResponse.Project;
 import org.sonarqube.ws.client.project.CreateRequest;
 
-import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
+import static joptsimple.internal.Strings.repeat;
```
[the remark this answers is not in the document]
Comment (gjambet) [reply server/sonar-server/src/test/java/org/sonar/server/project/ws/CreateActionTest.java]: right ! i mean no ! wrong ! but you're right, the import is wrong

File: server/sonar-server/src/test/java/org/sonar/server/project/ws/CreateActionTest.java [hunk 4]
```diff
@@ -287,6 +290,32 @@ public void fail_when_missing_create_project_permission() throws Exception {
   }
 
   @Test
+  public void fail_when_project_parameter_is_longer_than_400() {
```
[the remark this answers is not in the document]
Comment (gjambet) [reply server/sonar-server/src/test/java/org/sonar/server/project/ws/CreateActionTest.java]: test removed

File: server/sonar-server/src/test/java/org/sonar/server/project/ws/CreateActionTest.java [hunk 5]
```diff
@@ -287,6 +290,32 @@ public void fail_when_missing_create_project_permission() throws Exception {
...
+    userSession.addPermission(PROVISION_PROJECTS, db.getDefaultOrganization());
+
+    expectedException.expect(IllegalArgumentException.class);
+    expectedException.expectMessage("'project' length (401) is longer than the maximum authorized (400)");
```
[the remark this answers is not in the document]
Comment (gjambet) [reply server/sonar-server/src/test/java/org/sonar/server/project/ws/CreateActionTest.java]: no longer relevant

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java [hunk 6]
```diff
@@ -167,12 +162,42 @@ private String readParam(String key, @Nullable WebService.Param definition) {
...
+    checkArgument(valueLength <= maximumLength, "'%s' length (%s) is longer than the maximum authorized (%s)", key, valueLength, maximumLength);
+  }
+
+  private static void validateMaximumValue(String key, WebService.Param definition, String value) {
+    Integer maximumValue = definition.maximumValue();
+    if (maximumValue == null) {
+      return;
+    }
+    int valueAsInt = parseInt(value);
+    checkArgument(valueAsInt <= maximumValue, "'%s' parameter (value '%s') must be less than %s", key, valueAsInt, maximumValue);
```
[the remark this answers is not in the document]
Comment (gjambet) [reply sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java]: agreed

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java [hunk 7]
```diff
@@ -167,12 +162,42 @@ private String readParam(String key, @Nullable WebService.Param definition) {
...
+  }
+
+  private static int parseInt(String value) {
+    try {
+      return Integer.parseInt(value);
+    } catch (NumberFormatException exception) {
+      throw new IllegalStateException(format("The '%s' value cannot be parsed as an integer", value), exception);
```
[the remark this answers is not in the document]
Comment (gjambet) [reply sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java]: you're right it's more explicit

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ListAction.java#L133](https://github.com/SonarSource/sonarqube/blob/7698d6455713191e81dc9611c798c7a21218b12e/server/sonar-server/src/main/java/org/sonar/server/branch/ws/ListAction.java#L133): Specify a "BranchDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-web/src/main/js/api/metrics.ts [hunk 8]
```diff
@@ -21,8 +21,34 @@ import { getJSON } from '../helpers/request';
...
+  total: number;
+}
+
+export function getMetrics(data: { p?: number; ps?: number }): Promise<MetricsResponse> {
+  return getJSON('/api/metrics/search', data).catch(throwGlobalError);
+}
+
+export function getAllMetrics(
+  data: { p?: number; ps?: number } = { ps: 500 },
+  prev?: MetricsResponse
```
Comment (stas-vilchik) [inline server/sonar-web/src/main/js/api/metrics.ts]: I don't think it's a good idea to expose this parameter. Instead create an inner function:
```javascript
export function getAllMetrics(data) {
  return inner(data);

  function inner(data, prev) { ... }
}
```

File: server/sonar-web/src/main/js/api/metrics.ts [hunk 9]
```diff
@@ -21,8 +21,34 @@ import { getJSON } from '../helpers/request';
...
+): Promise<MetricsResponse> {
+  return getMetrics(data).then(r => {
```
Comment (stas-vilchik) [inline server/sonar-web/src/main/js/api/metrics.ts]: Since this function returns all metrics, there is no point to return paging information.

**Comment by gregaubert** [comment]
@stas-vilchik Fixed based on your comments

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ListAction.java#L133](https://github.com/SonarSource/sonarqube/blob/8811c36a54993a03e6eefe3517b7ac8ea6ea099e/server/sonar-server/src/main/java/org/sonar/server/branch/ws/ListAction.java#L133): Specify a "BranchDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/ws/ws/ListAction.java [hunk 10]
```diff
@@ -138,6 +138,10 @@ private static void writeParam(JsonWriter writer, WebService.Param param) {
     if (possibleValues != null) {
       writer.name("possibleValues").beginArray().values(possibleValues).endArray();
     }
+    Integer maximumLength = param.maximumLength();
```
[the remark this answers is not in the document]
Comment (julienlancelot) [reply server/sonar-server/src/main/java/org/sonar/server/ws/ws/ListAction.java]: It's not protobuf :(

**Pull request merged**